keps/sig-node/2625-cpumanager-policies-thread-placement/README.md

```diff
@@ -0,0 +1,314 @@
+
+Key properties:
+- Preserve all the properties of the `static` policy.
+- Never allocate less than a physical-cpu worth amount of cores.
+- With this requirement enforced, the cpumanager allocation algorithm will guarantee avoidance of physical core sharing.
+- Should the node not have enough free physical cores, the Pod will be put in Failed state, with `SMTAlignmentError` as reason.
+
+### Implementation strategy of smtaware CPU Manager policy
+
+- The CPU Manager implements the pod admit handler interface and participates in Kubelet pod admission.
+- The `GetAllocateResourcesPodAdmitHandler()` function in the Container Manager is modified to perform admission checks for CPU Manager in addition to the already occurring Topology manager admission check.
+Just like Topology Manager returns `TopologyAffinityError` is case of resources cannot be NUMA-aligned in case of `restricted` or `single-numa-node policy`, CPU Manager admission failure results in a rejected pod with `SMTAlignmentError`.
+- The Policy interface in CPU Manager is enhanced to support an Admit function. This allows to perform pod admission decision for CPU Manager policies.
+- When CPU Manager is configured with `smtaware` policy, when the `Admit()` function is called it is checked if the CPU request is
+such that it would acquire an entire physical core. In case request translates to partial occupancy of the cores the Pod will not be admitted and would fail with `SMTAlignmentError`. In case of `static` and `none` policy, the Admit() function always returns true meaning the pod is always admitted.
+- When `smtaware` policy is bootstrapped, we intentionally reuse `static` policy. The allocation logic of the `static` policy works seamlessly in case of `smtaware` policy as following the check at admission time of a pod, it is known that CPUs would be allocated such that full cores are allocated. Because of this check, a pod would never have to acquire single threads with the aim to fill partially-allocated cores.
+
+### Resource Accounting
+
+To illustrate the behaviour of the policy, we will consider the following CPU topology. We will use as example a CPU package with 16 physical cores, 2-way SMT-capable.
+
+![Example Topology](smtaware-topology.png)
+
+
+Let's consider a single container, requesting 5 isolated cores.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: frontend
+spec:
+  containers:
+  - name: app
+    image: images.my-company.example/app:v4
+    resources:
+      requests:
+        memory: "128Mi"
+        cpu: "5"
+      limits:
+        memory: "128Mi"
+        cpu: "5"
+```
+
+The `smtaware` policy would need to make sure the remaining core on the half-allocated physical CPU is left unallocated to avoid noisy neighbours.
+
+![Example core allocation with the smtaware policy when requesting a odd number of cores](smtaware-allocation-odd-cores.png)
+
+The container will then actually get more virtual cores (6) than what is requesting (5).
```

Comment:
The intent is to highlight the desired behaviour, and then show how the proposed implementation allows it.
The desired behaviour is that a pod requesting a odd number of CPUs is guaranteed that nothing else runs in the partially allocated physical core. Ideally
- from container perspective, 5 cores are requested, and 5 are given by the system
- from management perspective (kubelet) the container requests 5 cores, but the actual footprint of the container is 6, and the overallocation is needed to prevent anything else to run on the partially allocated core.

Reply:
> - from container perspective, 5 cores are requested, and 5 are given by the system
> - from management perspective (kubelet) the container requests 5 cores, but the actual footprint of the container is 6, and the overallocation is needed to prevent anything else to run on the partially allocated core.

I am just wondering since I looked at the PR in k8s/k8s, and in that implementation it looks like it will result in an error stating `Number of CPUs requested should be a multiple of number of CPUs on a core`. Or am I missing something?

https://github.com/kubernetes/kubernetes/pull/101432/files#diff-d9817f292659f2c017499f7e2578cf4ce7c2a48f266e239121e7363d78ec6ff9R1165-R1169

> - from management perspective (kubelet) the container requests 5 cores, but the actual footprint of the container is 6, and the overallocation is needed to prevent anything else to run on the partially allocated core.

This is exectly what I am wondering about. How is the idea of handling this overallocation? As I understand it, the pod-spec will say 5 cpus, while it will make sure no pod/container use the sibling core of the last one, essentially deducting 6 cpus from the node allocatable amount. This means that the k8s apiserver has to "know" that it should count as 6 cpus, or am I missing something?

Reply:
It seems you got it correctly indeed. We considered various options about how to represent the actual footprint of the container (6 allocated vs 5 requested) and no satisfying solution emerged. Representing a footprint bigger than a request (AFAICT) would request cascading changes all across kubernetes (random example: adding a `pod.Status.Request`, consuming all around the ecosystem) which are too large and very likely breaking. We had to settle about pushing this requirement out to the pod, which ensures all the resource allocation is consistently repored. We documented the discarded option (and most notably, why) in the KEP, please let me know if you think the provided rationale is not sufficient.

Reply:
Ahh, yeah, then it makes more sense! That should probably be clarified properly in the KEP tho., since now it looks like the "5 requested" cores on dual SMT will be supported. But from your comment (and the impl), it will just don't admit the pod.


> Representing a footprint bigger than a request (AFAICT) would cause cascading changes all across kubernetes (random example: adding a pod.Status.Request, consuming all around the ecosystem) which are too large and very likely breaking. 

Yes! That was my initial thought as well, and was something I do not think is a particular good idea right off the bat.

Reply:
ok, I'll review the KEP to make sure this is really evident.